compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -879,17 +893,73 @@ static void build_enum_cast_function(EnumType* et) {
                                  new_StringSymbol(errorString))));
     whenstmts->insertAtTail(otherwise);
     fn->insertAtTail(buildSelectStmt(new SymExpr(arg2), whenstmts));
-  }
-  DefExpr* def = new DefExpr(fn);
-  //
-  // these cast functions need to go in the base module because they
-  // are automatically inserted to handle implicit coercions
-  //
-  baseModule->block->insertAtTail(def);
-  reset_ast_loc(def, et->symbol);
-  normalize(fn);
 
-  fn->tagIfGeneric();
+    def = new DefExpr(fn);
+    baseModule->block->insertAtTail(def);
+    reset_ast_loc(def, et->symbol);
+    normalize(fn);
+    fn->tagIfGeneric();
+
+    // build the enumerated to integer cast
+    fn = new FnSymbol(astr_cast);
+    fn->addFlag(FLAG_COMPILER_GENERATED);
+    fn->addFlag(FLAG_LAST_RESORT);
+    arg1 = new ArgSymbol(INTENT_BLANK, "t", dtIntegral);
+    arg1->addFlag(FLAG_TYPE_VARIABLE);
+    arg2 = new ArgSymbol(INTENT_BLANK, "_arg2", et);
+    fn->insertFormalAtTail(arg1);
+    fn->insertFormalAtTail(arg2);
+
+    // Generate a select statement with when clauses for each of the
+    // enumeration constants, and an otherwise clause that calls halt.
+    count = 0;
+    whenstmts = buildChapelStmt();
+    lastInit = NULL;
+    for_enums(constant, et) {
+      if (constant->init) {
+        lastInit = constant->init;
+        count = 0;
+      } else {
+        if (lastInit != NULL) {
+          count++;
+        }
+      }
+      CallExpr* result;
+      if (lastInit) {
+        result = new CallExpr(PRIM_RETURN,
+                              new CallExpr(PRIM_CAST, arg1,
+                                           new CallExpr("+", lastInit->copy(),
+                                                        new SymExpr(new_IntSymbol(count)))));
+      } else {
+        const char *errorString = astr("illegal cast: ", et->symbol->name, 
+                                       ".", constant->sym->name,
+                                       " has no integer value");
+        result = new CallExpr("halt", new_StringSymbol(errorString));
+      }
+      CondStmt* when =
+        new CondStmt(new CallExpr(PRIM_WHEN, new SymExpr(constant->sym)),
+                     result);
+                                  
+      whenstmts->insertAtTail(when);
+    }
+
+    otherwise =
+      new CondStmt(new CallExpr(PRIM_WHEN),
+                   new BlockStmt(new CallExpr("halt",
+                                              new_StringSymbol(errorString))));
+    whenstmts->insertAtTail(otherwise);
+    fn->insertAtTail(buildSelectStmt(new SymExpr(arg2), whenstmts));
+
+    def = new DefExpr(fn);
+    //
+    // these cast functions need to go in the base module because they
+    // are automatically inserted to handle implicit coercions
```

Comment:
Is this comment still accurate? I would have thought they would no longer be used for implicit coercions?

Reply:
Good point.  We seem to run into problems whenever the compiler-provided enum functions aren't at this scope, so I'm updating the comment to avoid the mention of implicit coercions rather than moving where the functions are defined.